muser/src/main/java/edu/usf/sas/pal/muser/manager/MuserMusicDataAnalysisManager.java

```diff
@@ -0,0 +1,104 @@
+
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import edu.usf.sas.pal.muser.exception.FirebaseFileNotInitializedException;
+import edu.usf.sas.pal.muser.io.CSVFileWriter;
+import edu.usf.sas.pal.muser.io.FirebaseReader;
+import edu.usf.sas.pal.muser.model.MusicAnalysisModel;
+import edu.usf.sas.pal.muser.options.ProgramOptions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class MuserMusicDataAnalysisManager {
+
+    private CSVFileWriter csvFileWriter;
+
+    private FirebaseReader firebaseReader;
+
+    private ProgramOptions programOptions;
+
+    private List<MusicAnalysisModel> musicAnalysisList;
+
+    public MuserMusicDataAnalysisManager() throws FirebaseFileNotInitializedException {
+        firebaseReader = new FirebaseReader();
+        csvFileWriter = new CSVFileWriter();
+        musicAnalysisList = new ArrayList<>();
+        programOptions = ProgramOptions.getInstance();
+    }
+
+    public void processData() {
+        // create csv file and add the header
+        csvFileWriter.createHeader(MusicAnalysisModel.CSV_HEADER);
+        if (programOptions.getUserId() == null) {
+            // analyze all data and append the data in the csv file
+            analyzeAllUsersMusicData();
+        } else {
+            // analyze specific user data
+            processUserByIdForPlayerEvent(programOptions.getUserId());
+            processUserByIdForUiEvent(programOptions.getUserId());
+        }
+        //close the csv file
+        csvFileWriter.closeWriter();
+    }
+
+    private void analyzeAllUsersMusicData() {
+        List<QueryDocumentSnapshot> allUserIds = firebaseReader.getAllUserIds();
+        for (QueryDocumentSnapshot doc : allUserIds) {
+            processUserByIdForPlayerEvent(doc.getId());
+            processUserByIdForUiEvent(doc.getId());
+        }
+    }
+
+    private void processUserByIdForPlayerEvent(String userId) {
+        List<QueryDocumentSnapshot> userInfoById = new ArrayList<>(firebaseReader.getAllUserInfoById(userId));
+        for (QueryDocumentSnapshot doc : userInfoById) {
+            MusicAnalysisModel mam1 = new MusicAnalysisModel(userId, doc.getId());
+            MusicAnalysisModel mam = doc.toObject(MusicAnalysisModel.class);
+            processSetterForUiEventAndPlayerevenets(mam1, mam);
+        }
+    }
+
+    private void processUserByIdForUiEvent(String userId) {
+        List<QueryDocumentSnapshot> userInfoById = new ArrayList<>(firebaseReader.getAllUserUIInfoById(userId));
+        for (QueryDocumentSnapshot doc : userInfoById) {
+            MusicAnalysisModel mam1 = new MusicAnalysisModel(userId, doc.getId());
+            MusicAnalysisModel mam = doc.toObject(MusicAnalysisModel.class);
+            processSetterForUiEventAndPlayerevenets(mam1, mam);
+        }
+    }
+
+    private void processSetterForUiEventAndPlayerevenets(MusicAnalysisModel mam1, MusicAnalysisModel mam) {
+        mam1.setCurrentTimeMs(mam.getCurrentTimeMs());
```

Comment:
Instead of copying from one model instance to another, it would be far more efficient to just change the `MusicAnalysisModel` class to return the song variables directly when the getter is called.

For example, instead of:

`mam1.setmName(mam.song.getName());`

...you can change the implementation of `MusicAnalysisModel.getName()` to return `song.getName()`. That way you only have one copy of the variable in memory.

Then you can get rid of this method and just write the initial `mam` to the CSV file, instead of having `mam1` and `mam`.

Also, for any fields that are properties of the song, please prefix it with `song`. So the `MusicAnalysisModel.getName()` should be `MusicAnalysisModel.getSongName()`.